src/Akka.Hosting/ActorRegistry.cs

```diff
@@ -29,6 +67,16 @@ public class ActorRegistry : IExtension
         private readonly ConcurrentDictionary<Type, IActorRef> _actorRegistrations =
             new ConcurrentDictionary<Type, IActorRef>();
 
+        /// <summary>
+        /// In the event that an actor is not available yet, typically during the very beginning of ActorSystem startup,
+        /// we can wait on that actor becoming available.
+        /// </summary>
+        /// <remarks>
+        /// Have to store a collection of <see cref="WaitForActorRegistration"/>s here so each waiter gets its own cancellation token.
+        /// </remarks>
+        private readonly ConcurrentDictionary<Type, ImmutableHashSet<WaitForActorRegistration>> _actorWaiters =
```

Comment:
Contemplated just using a simple `lock` here but it became tricky once we started having the `CancellationTokenRegistration` delegates also competing to update entries. Opted for the compare-and-swap semantics of the `ConcurrentDictionary` instead.

> Man, this would have been so simple had I just encapsulated all of this inside an actor, but since we're trying to keep _this_ data structure from being dependent on the `ActorSystem` itself I went with this.